gen/src/download/mod.rs

```diff
@@ -8,107 +8,86 @@
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
 
+
+//! UNIC - Data Download Script
+//!
+//! # Config Format:
+//!
+//! ```toml
+//! [example]
+//! version = "10.0.0"
+//! base_url = "http://example.com/path/to/{version}/"
+//!
+//! [[example.resources]]
+//! "some_data.txt" = "some_data.txt"
+//!
+//! [[example.test.resources]]
+//! "other_data.txt" = "tests/other_data.txt"
+//! ```
+
+
 mod client;
 
+use std::collections::HashMap;
 use std::error::Error;
-use std::fs::{self, File};
-use std::io::Read;
-use std::path::Path;
+use std::path::PathBuf;
+use std::fs;
+
+use toml;
 
 use self::client::DownloadPath;
 
-use toml;
 
-/// The config file
-#[derive(Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
-struct Config {
-    idna: DataSource,
-    ucd: DataSource,
-    normal: DataSource,
-}
+type Config = HashMap<String, DataSource>;
 
-/// Information under one component heading in the config
-#[derive(Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
+#[derive(Debug, Deserialize)]
 struct DataSource {
     version: String,
-    resources: Vec<DownloadPath>,
+    base_url: String,
+    resources: HashMap<String, String>,
+    test_resources: HashMap<String, String>,
 }
 
-/// Path to configuration file.
-///
-/// Format:
-///
-/// ```toml
-/// [ucd]
-/// version = "10.0.0"
-///
-/// [[ucd.resources]]
-/// url = "http://example.com/resources/{version}/data.txt"
-/// dest = "data/example/data.txt"
-/// ```
-const DOWNLOADS: &'static str = "gen/config/config.toml";
 
-/// Downloads all files (indicated by `config/downloads.yaml`) for the indicated version.
-///
-/// Downloading the correct version is done via a naive string replacement from `{}` to the
-/// provided version string. Files are then saved in the indicated target location.
+/// Downloads all files from the config file.
 ///
 /// # Panics
 ///
-/// Panics if the config file is missing or badly formatted, or if it fails to open a file.
-/// Panics if an invalid component is passed in.
+/// Panics if the config file is badly formatted.
 ///
 /// # Errors
 ///
 /// Returns a boxed error when something goes wrong during download or a write error occurs.
-pub fn download(components: &[&str]) -> Result<(), Box<Error>> {
-    let mut file = File::open(Path::new(DOWNLOADS)).expect("Failed to open config.toml");
-    let mut buffer = String::new();
-    file.read_to_string(&mut buffer)
-        .expect("Failed to read config.toml");
-
-    let config: Config = toml::from_str(&buffer).expect("Failed to parse downloads.yaml");
+pub fn download(active_sources: &Vec<&str>) -> Result<(), Box<Error>> {
+    let config: Config =
+        toml::from_str(include_str!("config.toml")).expect("Failed to parse config file");
```

Comment:
Note: Using `include_str!` here means recompiling the module/crate when the config changes. That's the reason I originally used the file APIs to get at it, such that we could add more downloads without having to recompile.

It doesn't add much complexity to read the config file in at runtime rather than compile time, so I personally think its worth it.

Reply:
But I'm willing to go either way. Updating the config should be an infrequent enough thing that this doesn't matter.

Reply:
I think having less lines of code is a good goal here. Since running this tool is only useful in a dev environment, I don't see any benefit of not doing the simplest and shortest option. So, considering this and your comment, let's keep it then.